selectel/resource_selectel_iam_serviceuser_v1.go

```diff
@@ -0,0 +1,209 @@
+}
+
+func resourceIAMServiceUserV1Read(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	iamClient, diagErr := getIAMClient(meta)
+	if diagErr != nil {
+		return diagErr
+	}
+
+	log.Print(msgGet(objectServiceUser, d.Id()))
+	user, err := iamClient.ServiceUsers.Get(ctx, d.Id())
+	if err != nil {
+		if errors.Is(err, iamerrors.ErrUserNotFound) {
+			d.SetId("")
+			return nil
+		}
+
+		return diag.FromErr(errGettingObject(objectServiceUser, d.Id(), err))
+	}
+
+	d.Set("name", user.Name)
+	d.Set("enabled", user.Enabled)
+	d.Set("role", flattenIAMServiceUserRoles(user.Roles))
+	if _, ok := d.GetOk("password"); !ok {
+		d.Set("password", unknownFieldValue)
+	}
+
+	return nil
+}
+
+func resourceIAMServiceUserV1Update(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	iamClient, diagErr := getIAMClient(meta)
+	if diagErr != nil {
+		return diagErr
+	}
+
+	var password string
+	if d.Get("password") == unknownFieldValue {
+		password = ""
+	} else {
+		password = d.Get("password").(string)
+	}
+
+	_, err := iamClient.ServiceUsers.Update(ctx, d.Id(), serviceusers.UpdateRequest{
+		Enabled:  d.Get("enabled").(bool),
+		Name:     d.Get("name").(string),
+		Password: password,
+	})
+
+	if d.HasChange("role") {
+		oldState, newState := d.GetChange("role")
+		newRoles, err := getIAMServiceUserRolesFromSet(newState.(*schema.Set))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		oldRoles, err := getIAMServiceUserRolesFromSet(oldState.(*schema.Set))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		rolesToUnassign := make([]serviceusers.Role, 0)
+		rolesToAssign := make([]serviceusers.Role, 0)
+
+		for _, oldRole := range oldRoles {
+			if !slices.Contains(newRoles, oldRole) {
+				rolesToUnassign = append(rolesToUnassign, oldRole)
+			}
+		}
+
+		for _, newRole := range newRoles {
+			if !slices.Contains(oldRoles, newRole) {
+				rolesToAssign = append(rolesToAssign, newRole)
+			}
+		}
+		if len(rolesToAssign) != 0 {
+			err := iamClient.ServiceUsers.AssignRoles(ctx, d.Id(), rolesToAssign)
+			if err != nil {
+				return diag.FromErr(errUpdatingObject(objectServiceUser, d.Id(), err))
+			}
+		}
+
+		if len(rolesToUnassign) != 0 {
+			err := iamClient.ServiceUsers.UnassignRoles(ctx, d.Id(), rolesToUnassign)
+			if err != nil {
+				return diag.FromErr(errUpdatingObject(objectServiceUser, d.Id(), err))
+			}
+		}
+	}
+	if err != nil {
+		return diag.FromErr(errUpdatingObject(objectServiceUser, d.Id(), err))
+	}
+
+	return resourceIAMServiceUserV1Read(ctx, d, meta)
+}
+
+func resourceIAMServiceUserV1Delete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	iamClient, diagErr := getIAMClient(meta)
+	if diagErr != nil {
+		return diagErr
+	}
+
+	log.Print(msgDelete(objectServiceUser, d.Id()))
+	err := iamClient.ServiceUsers.Delete(ctx, d.Id())
+	if err != nil {
+		if errors.Is(err, iamerrors.ErrUserNotFound) {
+			d.SetId("")
```

Comment:
У нас не должно быть одинаковое поведение если пользователя уже нет и если мы его удалили? 

Если должно быть, то можно исправить на:
```suggestion
	if err != nil && !errors.Is(err, iamerrors.ErrUserNotFound){
		return diag.FromErr(errDeletingObject(objectServiceUser, d.Id(), err))
	}
	
	d.SetId("")
	return nil
```

`d.SetId("")` нужен при delete?

Reply:
Еща раз пересмотрел код, он, в целом, не нужен в `delete`-методах, поэтому `d.SetId("")` убрал совсем